apps/prairielearn/src/lib/authz-data.ts

```diff
@@ -1,156 +1,211 @@
  */
-export async function buildAuthzData({
-  authn_user,
+export async function calculateAuthData({
+  user,
   course_id,
   course_instance_id,
-  is_administrator,
   ip,
   req_date,
-  req_mode = null,
+  overrides,
 }: {
-  authn_user: User;
+  user: User;
   course_id: string | null;
   course_instance_id: string | null;
-  is_administrator: boolean;
   ip: string | null;
   req_date: Date;
-  req_mode?: string | null;
-}) {
+  overrides: {
+    is_administrator: boolean;
+    req_mode?: EnumMode;
+    req_course_role?: EnumCourseRole;
+    req_course_instance_role?: EnumCourseInstanceRole;
+    allow_example_course_override?: boolean;
+  };
+}): Promise<CalculateAuthDataResult> {
+  const resolvedOverrides = {
+    allow_example_course_override: true,
+    ...overrides,
+  };
+  assert(course_id !== null || course_instance_id !== null);
+
   const isCourseInstance = Boolean(course_instance_id);
 
   const rawAuthzData = await selectAuthzData({
-    user_id: authn_user.user_id,
+    user_id: user.user_id,
     course_id,
     course_instance_id,
-    is_administrator,
-    allow_example_course_override: true,
     ip,
     req_date,
-    req_mode,
-    req_course_role: null,
-    req_course_instance_role: null,
   });
 
   if (rawAuthzData === null) {
     return {
-      authzData: null,
-      authzCourse: null,
-      authzInstitution: null,
-      authzPermissionsCourse: null,
-      authzCourseInstance: null,
+      authResult: null,
+      course: null,
+      institution: null,
+      courseInstance: null,
+    };
+  }
+
+  const course_role = run(() => {
+    if (resolvedOverrides.req_course_role != null) {
+      return resolvedOverrides.req_course_role;
+    }
+    if (resolvedOverrides.is_administrator) {
+      return 'Owner';
+    }
+
+    if (rawAuthzData.course.example_course) {
+      if (resolvedOverrides.allow_example_course_override) {
+        return 'Viewer';
+      }
+      // If the course is an example course and the override is not allowed, return None.
+      return 'None';
+    }
+
+    return rawAuthzData.permissions_course.course_role;
+  });
+
+  const course_instance_role = run(() => {
+    if (resolvedOverrides.req_course_instance_role != null) {
+      return resolvedOverrides.req_course_instance_role;
+    }
+    if (resolvedOverrides.is_administrator) {
+      return 'Student Data Editor';
+    }
+    return rawAuthzData.permissions_course_instance.course_instance_role;
+  });
+
+  const mode = run(() => {
+    if (resolvedOverrides.req_mode != null) {
+      return resolvedOverrides.req_mode;
+    }
+    return rawAuthzData.mode;
+  });
+
+  // This block replicates the old behavior of selectAuthzData.
+  if (
+    course_role === 'None' &&
+    (isCourseInstance
+      ? course_instance_role === 'None' &&
+        !rawAuthzData.permissions_course_instance.has_student_access
+      : true)
```

Comment:
Nit, nested conditionals and ternaries can be a little tricky to read. Can this be refactored to something like:

```ts
const hasCourseAccess = course_role !== 'None';
const hasCourseInstanceAccess = course_instance_role !== 'None' || rawAuthzData.permissions_course_instance.has_student_access;
if (!hasCourseAccess && !hasCourseInstanceAccess) {
  ...
```

Reply:
Fixed in 3da233781605eef172362a8b171ee4d96550fd21.